agent/fake_api_server_test.go

```diff
@@ -0,0 +1,372 @@
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buildkite/agent/v3/api"
+	"github.com/google/uuid"
+)
+
+// This file implements a flexible fake testing server for the agent, including
+// registration through to pinging and running jobs, but not including many of
+// the other things the agent REST API does.
+//
+// Why not fake the client? Because there's a lot of value in testing that
+// real requests round-trip through a network connection, even if both endpoints
+// are the same process.
+
+const (
+	JobStateScheduled = "scheduled"
+	JobStateAssigned  = "assigned"
+	JobStateAccepted  = "accepted"
+	JobStateRunning   = "running"
+	JobStateFinished  = "finished"
+)
+
+type FakeJob struct {
+	Job   *api.Job
+	Auth  string
+	State string
+	Logs  strings.Builder
+}
+
+type FakeAgent struct {
+	Assigned   *FakeJob
+	Paused     bool
+	Stop       bool
+	Pings      int
+	Heartbeats int
+
+	PingHandler func() (api.Ping, error)
+}
+
+// agentJob is just an agent/job tuple.
+type agentJob struct {
+	agent *FakeAgent
+	job   *FakeJob
+}
+
+// FakeAPIServer implements a fake Agent REST API server for testing.
+type FakeAPIServer struct {
+	*httptest.Server
+
+	mu        sync.Mutex
+	agents    map[string]*FakeAgent // session token Auth header -> agent
+	jobs      map[string]*FakeJob   // uuid -> job
+	agentJobs map[string]agentJob   // job token Auth header -> (agent, job)
+}
+
+// NewFakeAPIServer constructs a new FakeAPIServer for testing.
+func NewFakeAPIServer() *FakeAPIServer {
+	fs := &FakeAPIServer{
+		agents:    make(map[string]*FakeAgent),
+		jobs:      make(map[string]*FakeJob),
+		agentJobs: make(map[string]agentJob),
+	}
+	mux := http.NewServeMux()
+	mux.HandleFunc("PUT /jobs/{job_uuid}/acquire", fs.handleJobAcquire)
+	mux.HandleFunc("PUT /jobs/{job_uuid}/accept", fs.handleJobAccept)
+	mux.HandleFunc("PUT /jobs/{job_uuid}/start", fs.handleJobStart)
+	mux.HandleFunc("PUT /jobs/{job_uuid}/finish", fs.handleJobFinish)
+	mux.HandleFunc("POST /jobs/{job_uuid}/chunks", fs.handleJobChunks)
+	mux.HandleFunc("GET /ping", fs.handlePing)
+	mux.HandleFunc("POST /heartbeat", fs.handleHeartbeat)
+	fs.Server = httptest.NewServer(mux)
+	return fs
+}
+
+func (fs *FakeAPIServer) AddAgent(token string) *FakeAgent {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	a := &FakeAgent{}
+	fs.agents["Token "+token] = a
+	return a
+}
+
+func (fs *FakeAPIServer) AddJob(env map[string]string) *FakeJob {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	jobToken := uuid.New().String()
+	j := &FakeJob{
+		State: JobStateScheduled,
+		Job: &api.Job{
+			ID:                 uuid.New().String(),
+			Token:              jobToken,
+			ChunksMaxSizeBytes: 1024,
+			Env:                env,
+		},
+		Auth: "Token " + jobToken,
+	}
+	fs.jobs[j.Job.ID] = j
+	return j
+}
+
+func (fs *FakeAPIServer) Assign(agent *FakeAgent, job *FakeJob) {
```

Comment:
Is this missing `fs.mu.Lock()` for potentially-concurrent `fs.agentJobs` map access?

Reply:
Yes it is!

Reply:
Would we perhaps use [`sync/v2`'s `Map`](https://github.com/golang/go/issues/71076) if/when that becomes a thing? Or is this mutex serving other purposes too?

Reply:
We could, which would potentially be a win from the perspectives of finer-grained locking and not having to remember to lock/unlock a mutex. The only downside would be potentially inconsistent views of the data, but this is for testing, so that's probably fine. (It might even be fine without a mutex at all.)

Reply:
> and not having to remember to lock/unlock a mutex

Yeah that's the bit I like, I don't have a strong instinct for mutexing things 😅